### DISCUSSION TITLE
ボールの有無でファームを分ける

### DISCUSSION THREAD

**Discussion by koron** [Yowkees/keyball#17] [Ideas]
いまボールの有無(基板の違い)によらずファームは1種類ですが、これを分けたほうが良いのではないかとのアイデアを得ました。思考の推移は以下の通りです。

* 左右自動判定やボールの有無の自動判定をやりたい
* ボールの有無の自動判定は何らかの方法でできそう (PMWにResetなげてレスポンスを見てみる)
* 左右の自動判定は無理そう

    ProMicroの未使用ピンがない & 異なる基板で未使用マトリックスが異なるため、1種のファームで2種の基板の左右を判定する(4通り!)ほどよい手段がない。ボールの有無を判定したのち、それに応じてチェックする未使用マトリックスを変えることはできるかもだが…複雑であまり好ましいアイデアとはいえない。

* ならいっそPCB毎に別のファームを用意したらどうか(名前の例: `rev1/onball`, `rev1/offball`)

    名前はあとで変更可

* これなら未使用マトリックスで左右判定できる
* ファームウェアのサイズ削減にもつながり一石二鳥
    * offballはPMWのファームがまるまる要らなくなる(大きな効果)
    * onballはトラックボールの有無判定が要らなくなる(小さな効果)
    * offballにだけviaを入れてUSBを繋ぐ(masterにする)ようにすると、onballからviaを消せてファームに余裕ができるのでは?

        VIAはUSBを繋いでるほうだけで良いはず。TRRSの通信内容はマトリクスの転送やRGB LEDのタイミング同期だけでVIA固有のものはない

* 判定に使うマトリクスは onball が `F7, B5` で offball が `F6, B5` を予定
    * これはonballが右手の default キーマップでは MBTN1 と MBTN2 に相当
* `rev1` は引き続き自動判定なし、トラックボール側にUSBを接続する&左手ボールの場合は適切なファームを使う、で使える

**Comment by Yowkees** [comment]
絶対このアイデアが正解だと思います！
現状右利き用と左利き用の2種のファームウェアを準備する必要がありますが、
それをボールありなしに分けることで、メモリの削減もできるわけですね。

ちなみに、この改造を行うとonballファームウェアを左右に書き込めばダブルボール仕様が実現してしまうのでしょうか！？

**Reply by koron** [reply]
> この改造を行うとonballファームウェアを左右に書き込めばダブルボール仕様が実現してしまうのでしょうか！？

ですですw VIAは使えなくなるかもですが、ダブルボールを見据えたアイデアです。

**Reply by Yowkees** [reply]
すごい！すごい！
ダブルボールは理論上可能とは思っていましたが、実際に使用感を確認できるのは半年以上先だろうなぁ〜と思っていました！😆ぜひ使ってみたいです！！

**Comment by koron** [comment]
@Yowkees 

> * ならいっそPCB毎に別のファームを用意したらどうか(名前の例: rev1/onball, rev1/offball)
> 
>     名前はあとで変更可

これなのですが何か名前の希望はありますか? オススメなのは rev1/ 下に置くことなのですが
rev1ball/, rev1noball/ みたいに rev1/ の横に置くこともできます。

ビルド時のコマンドがこんな感じになるイメージです。

make keyball/rev1/onball:default

**Reply by Yowkees** [reply]
おすすめいただきました通りrev1/ 下に置きたいです。
ball と noball の切り換えが僕には理解しやすかったので、これでどうでしょうか！？

いろいろ考えていただいて、本当にありがとうございます！(TOT)
